### PR TITLE
Incidence Angle Modifier zero instead of np.nan (#338)

### PR DESCRIPTION
**Pull request by jkfm**
This PR changes the three IAM-models to return 0 instead of np.nan for aoi >= 90 (ashraeiam, physicaliam) and <0 (sapm_aoi_loss). 

I also updated the tests to account for the changes. The test for physicaliam now shows, that angles of -90 and 90 return 0, whereas an angle of 0 returns 1, in accordance to the model description in De Soto et al..

Comments are welcome!

### CODE REVIEW

**Comment by adriesse** [comment]
Could we have the identical code in each of the three functions to do the same thing?  That is, to deal with the negative values and the larger than 90 angles?  I don't know whether np.maximum, np.where or np.clip is better, but I would lean toward clip.

The SAPM model doesn't like negative angles, the others are fine because of the cosine.  That should be handled consistently as well.

File: pvlib/pvsystem.py [hunk 1]
```diff
@@ -811,6 +812,10 @@ def physicaliam(aoi, n=1.526, K=4., L=0.002):
     spa
     ashraeiam
     '''
+    zeroang = 1e-06
+
+    aoi = np.where(aoi == 0, zeroang, aoi)
```
Comment (wholmgren) [inline pvlib/pvsystem.py]: Why did you add this line? I understand what it does, but not why we need it now when we didn't previously.

It's not clear to me that we actually need zeroang at all. I think its only purpose is to avoid infs, but maybe those be more cleanly handled with np.where or similar.

File: pvlib/pvsystem.py [hunk 2]
```diff
@@ -732,7 +732,7 @@ def ashraeiam(aoi, b=0.05):
 
     iam = 1 - b*((1/np.cos(np.radians(aoi)) - 1))
 
-    iam = np.where(np.abs(aoi) >= 90, np.nan, iam)
+    iam = np.where(np.abs(aoi) >= 90, 0, iam)
     iam = np.maximum(0, iam)
```
Comment (wholmgren) [inline pvlib/pvsystem.py]: is the np.maximum line still necessary?

**Comment by wholmgren** [comment]
@adriesse brings up a separate but closely related point regarding the meaning of negative values of aoi in pvlib. I tend to agree that the aoi loss functions should all start with something like ``aoi = np.abs(aoi)``, but that's open for discussion. I thought that we had previously discussed this elsewhere, but I can't find the discussion. If we're going to change it, now is the time.

**Comment by cwhanse** [comment]
If there are negative values for AOI, that indicates a problem elsewhere.  I’m not in favor of using abs(aoi) to address it.

File: pvlib/pvsystem.py (hunk 2 above)
Comment (jkfm) [inline pvlib/pvsystem.py]: I think it is, because for angles close to +/-90° the function would return a negative value for the iam (the intercept for b=0.05 is at 87.21°: https://www.desmos.com/calculator/y8wvzrx3b9).

File: pvlib/pvsystem.py (hunk 1 above)
Comment (jkfm) [inline pvlib/pvsystem.py]: As I understood it, if the input angle is 0, tau will be nan. 
zeroang = 1e-06 is defined for all cases, because iam is defined as tau(theta) / tau(0°), so that tau(0°) has to be calculated anyway. I didn't actually add the line, but moved it up from further down.

I guess we could define tau0 as a constant in the function and handling aoi == 0 with np.where after the calculations are done.

Comment (adriesse) [inline pvlib/pvsystem.py]: I am surprised that the model would not work for an angle of exactly zero degrees.  Could this have something to do with the equation discrepancy mentioned in the comments further above?

**Comment by cwhanse** [comment]
I vote for IAM=0 where AOI>=90, which is consistent with its definition and its use in the modeling process.

Where AOI=0 (or very close to 0) in the physical IAM model the formula needs to switch to the limit expression in my earlier comment. That switch should be the use of zeroang.

Sent from my iPhone

On Jun 30, 2017, at 3:20 PM, Anton Driesse <notifications@github.com<mailto:notifications@github.com>> wrote:


@adriesse commented on this pull request.

________________________________

In pvlib/pvsystem.py<https://github.com/pvlib/pvlib-python/pull/339#discussion_r125122764>:

> @@ -811,6 +812,10 @@ def physicaliam(aoi, n=1.526, K=4., L=0.002):
     spa
     ashraeiam
     '''
+    zeroang = 1e-06
+
+    aoi = np.where(aoi == 0, zeroang, aoi)


I am surprised that the model would not work for an angle of exactly zero degrees. Could this have something to do with the equation discrepancy mentioned in the comments further above?

-

File: pvlib/pvsystem.py (hunk 1 above)
Comment (jkfm) [inline pvlib/pvsystem.py]: I don't think it has to do with the function discrepancy. The unmodified function for theta_r is not defined for angles > ~40°.

The tau(theta)-function in turn is just not defined for theta = 0. It seems this was ignored in the paper...

**Comment by jkfm** [comment]
I would also prefer IAM to return zero for angles outside -90/90 degrees.

File: pvlib/pvsystem.py [hunk 3]
```diff
@@ -831,7 +834,7 @@ def physicaliam(aoi, n=1.526, K=4., L=0.002):
```
Comment (adriesse) [inline pvlib/pvsystem.py]: I agree with your suggestions, especially after having consulted my D&B book.  I would suggest calculating reflectance and absorption separately before combining them in order to clarify the code.

Comment (jkfm) [inline pvlib/pvsystem.py]: I think I applied both your suggestions correctly. Could you take a look?

Comment (adriesse) [inline pvlib/pvsystem.py]: It wasn't quite correct, so I thought it might be easier to communicate it this way:
```
    # angle of refraction
    thetar_deg = tools.asind(tools.sind(aoi) / n)

    # reflectance and transmittance for normal incidence light
    rho_zero = ((1-n) / (1+n)) ** 2
    tau_zero = np.exp(-K*L)

    # reflectance for parallel and perpendicular polarized light
    rho_para = (tools.tand(thetar_deg - aoi) /
                tools.tand(thetar_deg + aoi)) ** 2
    rho_perp = (tools.sind(thetar_deg - aoi) /
                tools.sind(thetar_deg + aoi)) ** 2

    # transmittance for non-normal light
    tau = np.exp(-K*L / tools.cosd(thetar_deg))

    # iam is ratio of non-normal to normal incidence transmitted light
    # after deducting the reflected portion of each
    iam = ((1 - (rho_para + rho_perp) / 2) / (1 - rho_zero)
          * tau / tau_zero )


    # angles near zero produce nan, but iam is defined as one
    small_angle = 1e-06
    iam = np.where(np.abs(aoi) < small_angle, 1.0, iam)

    # angles at 90 degrees can produce tiny negative values, which should be zero
    # this is a result of calculation precision rather than the physical model
    iam = np.where(iam <  0, 0, iam)    

    # for light coming from behind the plane, none can enter the module
    iam = np.where(aoi > 90, 0, iam)

    # negative angles are not logical
    iam = np.where(aoi <  0, np.nan, iam)

```

Comment (jkfm) [inline pvlib/pvsystem.py]: Thanks for the detailed explanation! So does this refer to reference 2 in the Docstring (I don't have access so cannot check)?

Comment (adriesse) [inline pvlib/pvsystem.py]: Ref [1] simply describes what is in ref [2].

**Comment by adriesse** [comment]
Another observation: contrary to the statement in ref [1] the default value of 2 mm for the glass thickness is definitely at the thin extreme.  The value I currently see most often is 3.2 mm. Previously I believe 4 mm was more common.

**Comment by jkfm** [comment]
I agree with the default value, but I could not find any sources on it. Do you think, we should change the default nevertheless?

File: pvlib/pvsystem.py (hunk 3 above)
Comment (jkfm) [inline pvlib/pvsystem.py]: I implemented your suggestion except for setting negative angles to np.nan, to keep consistency with the other IAM functions for now...

Comment (adriesse) [inline pvlib/pvsystem.py]: Consistency has some appeal...

**Comment by adriesse** [comment]
The source for the 2mm default has no sources either...

**Comment by cwhanse** [comment]
If it helps, the model output is essentially insensitive to the thickness parameter.  You can double from 0.002 to 0.004 and the IAM changes very slightly.

For myself, I'd prefer the default to remain the value in the reference paper, simply for traceability back to literature.  I do see merit in changing to 3.2mm, though.

**Comment by wholmgren** [comment]
I suggest keeping the default set to 2 mm. I'll merge tomorrow unless there are objections.
